proposals/p5337.md

```diff
@@ -800,9 +801,40 @@ fn H(V:! W) {
 }
 ```
 
-The compiler will only use a known matching final `impl`'s associated constants
-if it is the first that _could_ match, as determined by type structure. For
-example:
+> Note: `final match_first` was considered in
+> [open discussion on 2023-09-13](https://docs.google.com/document/d/1s3mMCupmuSpWOFJGnvjoElcBIe2aoaysTIdyczvKX84/edit?resourcekey=0-G095Wc3sR6pW1hLJbGgE0g&tab=t.0#heading=h.4r37mjr7c42h)
+> as a part of function overloading, and other times since.
+
+This leads us to these rules:
+
+-   An `impl` is considered "final" if is declared with the `final` keyword
+    modifier or if it is named in a `final match_first` block.
+    -   In the first case (a `final impl` declaration), that `impl` may not
+        appear in any `match_first` block.
+-   An `impl` may appear in at most one `match_first` block.
+    -   This is true whether or not the block is marked `final`.
+-   Two final `impl`s in the same file must appear in the same
+    `final match_first` block if they overlap, as determined by their type
+    structure.
```

Comment:
I had to read back through the proposal carefully to find that we also only allow final impls that overlap in a single file, for now. Maybe we can say that here?